From: intrigeri <intrigeri@boum.org>
Date: Mon, 30 Jan 2012 18:29:47 +0100
Subject: If a symbol cannot be found, also try by prefixing its name with __.

Rationale: libresolv is considered as a private interface of eglibc.
Private symbols in there see their names prefixed this way.
Like it or not, we have to go find them using the names they go by with.

Trying both kind of names maximizes our support of various vendors of the libc
and various versions of the glibc (FTR, symbols in there where renamed between
2.0 and 2.3).
---
 src/expansion_table.h |    5 -----
 src/torsocks.c        |   20 +++++++++++++-------
 2 files changed, 13 insertions(+), 12 deletions(-)

diff --git a/src/expansion_table.h b/src/expansion_table.h
index e981f5d..14fabe1 100644
--- a/src/expansion_table.h
+++ b/src/expansion_table.h
@@ -76,12 +76,7 @@
 /*RES_FUNC  (ERR,    int,                RES_INIT_,          res_init,                      res_init,            "res_init") */
 RES_FUNC    (ERR,    int,                RES_QUERY_,         res_query,                     res_query,           "res_query")
 RES_FUNC    (ERR,    int,                RES_SEARCH_,        res_search,                    res_search,          "res_search")
-#if defined(__APPLE__) || defined(__darwin__)
 RES_FUNC    (ERR,    int,                RES_SEND_,          res_send,                      res_send,            "res_send")
-#else
-/* It is a bit of a mystery why this is required on Linux. See http://code.google.com/p/torsocks/issues/detail?id=3 */
-RES_FUNC    (ERR,    int,                RES_SEND_,          res_send,                      res_send,            "__res_send")
-#endif
 RES_FUNC    (ERR,    int,                RES_QUERYDOMAIN_,   res_querydomain,               res_querydomain,     "res_querydomain")
 
 DNS_FUNC    (ERR,    struct hostent *,   GETHOSTBYNAME_,     gethostbyname,                 gethostbyname,       "gethostbyname")
diff --git a/src/torsocks.c b/src/torsocks.c
index f559eec..fcd9d95 100644
--- a/src/torsocks.c
+++ b/src/torsocks.c
@@ -150,10 +150,11 @@ void torsocks_init(void)
     dlerror();
 #ifndef USE_OLD_DLSYM
     #ifdef SUPPORT_RES_API
-    if ((realres_init = dlsym(RTLD_NEXT, "res_init")) == NULL)
+    if (((realres_init = dlsym(RTLD_NEXT, "res_init")) == NULL) &&
+        ((realres_init = dlsym(RTLD_NEXT, "__res_init")) == NULL))
         LOAD_ERROR("res_init", MSGERR);
     #endif
-    #define PATCH_TABLE_EXPANSION(e,r,s,n,b,m)  if ((real##n = dlsym(RTLD_NEXT, m)) == NULL) LOAD_ERROR(m, MSG##e);
+    #define PATCH_TABLE_EXPANSION(e,r,s,n,b,m)  if (((real##n = dlsym(RTLD_NEXT, m)) == NULL) && ((real##n = dlsym(RTLD_NEXT, "__" m)) == NULL)) LOAD_ERROR(m, MSG##e);
     #include "expansion_table.h"
     #undef PATCH_TABLE_EXPANSION
 #else
@@ -856,7 +857,8 @@ int res_init(void)
 {
     int rc;
 
-    if (!realres_init && ((realres_init = dlsym(RTLD_NEXT, "res_init")) == NULL))
+    if (!realres_init && ((realres_init = dlsym(RTLD_NEXT, "res_init")) == NULL) &&
+                         ((realres_init = dlsym(RTLD_NEXT, "__res_init")) == NULL))
         LOAD_ERROR("res_init", MSGERR);
 
     show_msg(MSGTEST, "Got res_init request\n");
@@ -877,7 +879,8 @@ int EXPAND_GUTS_NAME(res_query)(RES_QUERY_SIGNATURE, int (*original_res_query)(R
 {
     int rc;
 
-    if (!original_res_query && ((original_res_query = dlsym(RTLD_NEXT, "res_query")) == NULL))
+    if (!original_res_query && ((original_res_query = dlsym(RTLD_NEXT, "res_query")) == NULL) &&
+                               ((original_res_query = dlsym(RTLD_NEXT, "__res_query")) == NULL))
         LOAD_ERROR("res_query", MSGERR);
 
     show_msg(MSGTEST, "Got res_query request\n");
@@ -903,7 +906,8 @@ int EXPAND_GUTS_NAME(res_querydomain)(RES_QUERYDOMAIN_SIGNATURE, int (*original_
     int rc;
 
     if (!original_res_querydomain &&
-        ((original_res_querydomain = dlsym(RTLD_NEXT, "res_querydomain")) == NULL))
+        ((original_res_querydomain = dlsym(RTLD_NEXT, "res_querydomain")) == NULL) &&
+        ((original_res_querydomain = dlsym(RTLD_NEXT, "__res_querydomain")) == NULL))
         LOAD_ERROR("res_querydoimain", MSGERR);
 
     show_msg(MSGDEBUG, "Got res_querydomain request\n");
@@ -929,7 +933,8 @@ int EXPAND_GUTS_NAME(res_search)(RES_SEARCH_SIGNATURE, int (*original_res_search
     int rc;
 
     if (!original_res_search &&
-        ((original_res_search = dlsym(RTLD_NEXT, "res_search")) == NULL))
+        ((original_res_search = dlsym(RTLD_NEXT, "res_search")) == NULL) &&
+        ((original_res_search = dlsym(RTLD_NEXT, "__res_search")) == NULL))
             LOAD_ERROR("res_search", MSGERR);
 
     show_msg(MSGTEST, "Got res_search request\n");
@@ -954,7 +959,8 @@ int EXPAND_GUTS_NAME(res_send)(RES_SEND_SIGNATURE, int (*original_res_send)(RES_
 {
     int rc;
 
-    if (!original_res_send && ((original_res_send = dlsym(RTLD_NEXT, "res_send")) == NULL))
+    if (!original_res_send && ((original_res_send = dlsym(RTLD_NEXT, "res_send")) == NULL)
+                           && ((original_res_send = dlsym(RTLD_NEXT, "__res_send")) == NULL))
             LOAD_ERROR("res_send", MSGERR);
 
     show_msg(MSGTEST, "Got res_send request\n");
-- 
